text/0000-allow-self-in-where-clauses.md

```diff
@@ -0,0 +1,75 @@
+- Feature Name: `allow_self_in_where_clauses`
+- Start Date: 2016-06-13
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes allowing the `Self` type to be used in where clauses for trait
+implementations, as well as referencing associated types for the trait being
+implemented.
+
+# Motivation
+[motivation]: #motivation
+
+`Self` is a useful tool to have to reduce churn when the type changes for
+various reasons. One would expect to be able to write
+
+```rust
+impl SomeTrait for MySuperLongType<T, U, V, W, X> where
+  Self: SomeOtherTrait,
+```
+
+but this will fail to compile today, forcing you to repeat the type, and adding
+one more place that has to change if the type ever changes.
+
+By this same logic, we would also like to be able to reference associated types
+from the traits being implemented. When dealing with generic code, patterns like
+this often emerge:
+
+```rust
+trait MyTrait {
+    type MyType: SomeBound;
+}
+
+impl<T, U, V> MyTrait for SomeStruct<T, U, V> where
+    SomeOtherStruct<T, U, V>: SomeBound,
```

Comment:
If you could change this from `SomeOtherStruct<T, U, V>` to `<Self as MyTrait>::MyType`, that would make clear the current verbosity is not quite as bad :).